[inductor] Reduce duplication of reduction combine functions

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102737
* #102486
* #102736
* __->__ #99661

Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that inplace reduction operations are now written as
an out-of-place operation and an assignment, e.g. instead if
```cpp
omp_out += omp_in
```
we generate
```cpp
omp_out = omp_out + omp_in
```

Which is a purely cosmetic change

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

peterbell10 committed:
[inductor] Make combine_fn an argument of ir.Reduce

This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Fix unrolled argmin/argmax on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Revert tl.full usage on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make combine_fn an argument of ir.Reduce"


This is attempting to make `ir.Reduce` and `ops.reduction` more
generic, so the reduction doesn't need to be hard-coded separately in
`ir.py`, `codegen/triton.py` and `codegen/cpp.py`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Share reduction combine_fn between backends"


Currently reductions are hard-coded separately in `ir.py`, `codegen/triton.py`
and `codegen/cpp.py`. This reduces some of that duplication by factoring out the
`combine_fn` definition used in `_unroll_reduction_fn` and using it to generate
some of the reduction code for the triton and cpp backends.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Share reduction combine_fn between backends"


Currently reductions are hard-coded separately in `ir.py`, `codegen/triton.py`
and `codegen/cpp.py`. This reduces some of that duplication by factoring out the
`combine_fn` definition used in `_unroll_reduction_fn` and using it to generate
some of the reduction code for the triton and cpp backends.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Workaround tl.reduce not working for int1 on "[inductor] Share reduction combine_fn between backends"


Currently reductions are hard-coded separately in `ir.py`, `codegen/triton.py`
and `codegen/cpp.py`. This reduces some of that duplication by factoring out the
`combine_fn` definition used in `_unroll_reduction_fn` and using it to generate
some of the reduction code for the triton and cpp backends.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Share reduction combine_fn between backends"


Currently reductions are hard-coded separately in `ir.py`, `codegen/triton.py`
and `codegen/cpp.py`. This reduces some of that duplication by factoring out the
`combine_fn` definition used in `_unroll_reduction_fn` and using it to generate
some of the reduction code for the triton and cpp backends.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Share reduction combine_fn between backends"


Currently reductions are hard-coded separately in `ir.py`, `codegen/triton.py`
and `codegen/cpp.py`. This reduces some of that duplication by factoring out the
`combine_fn` definition used in `_unroll_reduction_fn` and using it to generate
some of the reduction code for the triton and cpp backends.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Reduce duplication of reduction combine functions"


Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that `vec_reduce_all` will do an
unnecessary mutation within the lambda, e.g. instead if
```cpp
[](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>&y) {return at::vec::maximum(x, y);}
```
we generate
```cpp
[](at::vec::Vectorized<float> x, at::vec::Vectorized<float> y) {return x = at::vec::maximum(x, y);}
```

The extra mutation is benign since the arguments are no longer a reference.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "[inductor] Reduce duplication of reduction combine functions"


Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that `vec_reduce_all` will do an
unnecessary mutation within the lambda, e.g. instead if
```cpp
[](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>&y) {return at::vec::maximum(x, y);}
```
we generate
```cpp
[](at::vec::Vectorized<float> x, at::vec::Vectorized<float> y) {return x = at::vec::maximum(x, y);}
```

The extra mutation is benign since the arguments are no longer a reference.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Reduce duplication of reduction combine functions"


Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that `vec_reduce_all` will do an
unnecessary mutation within the lambda, e.g. instead if
```cpp
[](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>&y) {return at::vec::maximum(x, y);}
```
we generate
```cpp
[](at::vec::Vectorized<float> x, at::vec::Vectorized<float> y) {return x = at::vec::maximum(x, y);}
```

The extra mutation is benign since the arguments are no longer a reference.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Reduce duplication of reduction combine functions"


Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that `vec_reduce_all` will do an
unnecessary mutation within the lambda, e.g. instead if
```cpp
[](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>&y) {return at::vec::maximum(x, y);}
```
we generate
```cpp
[](at::vec::Vectorized<float> x, at::vec::Vectorized<float> y) {return x = at::vec::maximum(x, y);}
```

The extra mutation is benign since the arguments are no longer a reference.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Reduce duplication of reduction combine functions"


Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that `vec_reduce_all` will do an
unnecessary mutation within the lambda, e.g. instead if
```cpp
[](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>&y) {return at::vec::maximum(x, y);}
```
we generate
```cpp
[](at::vec::Vectorized<float> x, at::vec::Vectorized<float> y) {return x = at::vec::maximum(x, y);}
```

The extra mutation is benign since the arguments are no longer a reference.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Reduce duplication of reduction combine functions"


Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that `vec_reduce_all` will do an
unnecessary mutation within the lambda, e.g. instead if
```cpp
[](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>&y) {return at::vec::maximum(x, y);}
```
we generate
```cpp
[](at::vec::Vectorized<float> x, at::vec::Vectorized<float> y) {return x = at::vec::maximum(x, y);}
```

The extra mutation is benign since the arguments are no longer a reference.

[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1327,19 +1327,8 @@ def final_argreduce(buffer, result_var, value, index):
-
+                combine_fn = ir.get_reduction_combine_fn(reduction_type, src_dtype)
+                updated = combine_fn(accumulator, value)
                 self.compute.writeline(
                     f"{accumulator} = tl.where({cond}, {updated}, {accumulator})"
                 )
```
Idea, but not for this PR. If we could know that `value` has just been loaded, we could elide this `tl.where`, as in those cases, the `other=` kwarg within `tl.load` already sets those extra values to the correct identity element.

I don't think we have a nice way of doing this now though, as the CSE vars carry close to no information.

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1327,19 +1327,8 @@ def final_argreduce(buffer, result_var, value, index):
-                else:
-                    raise NotImplementedError(f"reduction_type {reduction_type}")
-
+                combine_fn = ir.get_reduction_combine_fn(reduction_type, src_dtype)
```
`arg{max,min}` do not go through this path, and rather re-implement everything. Any chance we can merge those two paths as well?

peterbell10 replied:
tbh I think `triton_helpers.minimum_with_index` is a lot more readable than generating the whole combine function.

lezcano replied:
That's right. This was more of a

lezcano replied:
That's right. The only thing is that, for that path, we still have two implementations of the same thing, but I guess that's alright.

lezcano replied:
It'd be good to have a note asking people to keep these two in sync if they modify one or the other. No need to do it in this PR tho.

lezcano reviewed (commented):
To generate C++ code that is a bit less cryptic, it'd be good to have an `inplace=False` kwarg on `reduction_combine_vec`.

lezcano reviewed (commented):
To generate C++ code that is a bit less cryptic, it'd be good to have an `inplace=False` kwarg on `reduction_combine_vec`.

peterbell10 committed:
Update on "[inductor] Reduce duplication of reduction combine functions"


Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that `vec_reduce_all` will do an
unnecessary mutation within the lambda, e.g. instead if
```cpp
[](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>&y) {return at::vec::maximum(x, y);}
```
we generate
```cpp
[](at::vec::Vectorized<float> x, at::vec::Vectorized<float> y) {return x = at::vec::maximum(x, y);}
```

The extra mutation is benign since the arguments are no longer a reference.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Reduce duplication of reduction combine functions"


Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that `vec_reduce_all` will do an
unnecessary mutation within the lambda, e.g. instead if
```cpp
[](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>&y) {return at::vec::maximum(x, y);}
```
we generate
```cpp
[](at::vec::Vectorized<float> x, at::vec::Vectorized<float> y) {return x = at::vec::maximum(x, y);}
```

The extra mutation is benign since the arguments are no longer a reference.

[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1403,28 +1396,20 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
+            )
             vec_reduce_all_func = f"{vec_ns}::vec_reduce_all<{DTYPE_TO_CPP[dtype]}>"
-            next_value = f"{vec_reduce_all_func}([]({vec}& x, {vec}&y) {reduce_all_body}, {tmpvar_vec})"
+            next_value = f"{vec_reduce_all_func}([]({vec}& x, {vec}& y) {reduce_all_body}, {tmpvar_vec})"
```
Nit. The previous change of passing these by value was potentially beneficial. As these variables fit in registers, they are better passed by value. Perhaps the compiler is able to optimise out this lambda though.

peterbell10 committed:
Update on "[inductor] Reduce duplication of reduction combine functions"


Currently reduction bodies are duplicated in several different places.
This reduces duplication by `combine_fn` definition used in
`_unroll_reduction_fn` and using it in the triton codegen. For cpp
this also makes better use of `reduction_combine{,_vec}` by using them
to generate the `omp declare reduction` line and the `vec_reduce_all`
call.

For triton the only change is that that the combine step gets spread
over two lines, e.g. instead of:
```python
_tmp1 = tl.where(rmask & xmask, triton_helpers.maximum(_tmp1, tmp0), _tmp1)
```
we get
```python
tmp2 = triton_helpers.maximum(_tmp1, tmp0)
_tmp1 = tl.where(rmask & xmask, tmp2, _tmp1)
```

For cpp the only change is that inplace reduction operations are now written as
an out-of-place operation and an assignment, e.g. instead if
```cpp
omp_out += omp_in
```
we generate
```cpp
omp_out = omp_out + omp_in
```

Which is a purely cosmetic change

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
